src/charm.py

```diff
@@ -205,6 +207,110 @@ def _check_kf_profiles(self, interfaces):
 
         return kf_profiles
 
+    def main(self, event) -> None:
+        """Main entry point for the Charm."""
+        try:
+            self._check_container_connection()
+            self._check_model_name()
+            self._check_leader()
+            interfaces = self._get_interfaces()
+            kf_profiles = self._check_kf_profiles(interfaces)
+            self.handle_ingress(interfaces)
+        except CheckFailed as e:
+            self.model.unit.status = e.status
+            return
+        try:
+            self.unit.status = MaintenanceStatus("Creating k8s resources")
+            self.k8s_resource_handler.apply()
+        except ApiError:
+            self.logger.error(traceback.format_exc())
+            self.unit.status = BlockedStatus("kubernetes resource creation failed")
+            return
+        self.handle_ingress(interfaces)
+        kf_profiles = list(kf_profiles.get_data().values())[0]
+        self.profiles_service = kf_profiles["service-name"]
+        try:
+            self._update_layer()
+        except CheckFailed as e:
+            self.model.unit.status = e.status
+            return
+        self.model.unit.status = ActiveStatus()
+
+    def _get_sidebar_config(self) -> Dict:
+        try:
+            current_configmap = self.k8s_resource_handler.lightkube_client.get(
+                ConfigMap, name=self._context["configmap_name"]
+            )
+            old_sidebar_config = json.loads(current_configmap.data["links"])
+        except ApiError as e:
+            self.logger.warning(
+                f"PROBLEM during configmap retrieval {e}. USING BASE CONFIG"
+            )
+            old_sidebar_config = json.loads(BASE_SIDEBAR)
+        return old_sidebar_config
+
+    def _update_sidebar_config(self, links_data: Dict) -> None:
+        self._context["links"] = json.dumps(links_data)
+        try:
+            self.k8s_resource_handler.context = self._context
+            self.k8s_resource_handler.template_files = [
+                self._resource_files["config_maps"]
+            ]
+            self.k8s_resource_handler.apply()
+        except ApiError as e:
+            self.unit.status = BlockedStatus(
+                f"kubernetes resource creation failed with {e}"
+            )
+
+    def _on_sidebar_relation_changed(self, event: RelationChangedEvent) -> None:
```

Comment:
I wonder if we could make the sidebar code more robust and simpler by making it more declarative.  Instead of worrying about adding/removing links on events, we could just recreate the sidebar whenever something happens that might affect it.  Sidebar recomputation is cheap, so afaict there isn't a real downside.

What if, on any event that can affect the sidebar (config_changed, sidebar_relation_*, ...), the charm:
* recomputes the full sidebar from the current relations
* `if new configmap!=old configmap`, push the changes

That can be included in `main()`, which will also help handling things like config_changed events as noted elsewhere. 
 We could optionally make a handler for `on_sidebar_relation_*` events that is simpler than `main`, but just running `main` on all events should work without much downside.  It also simplifies the charm's logic and I think will need maybe 1/3 the code.

A side benefit of this is that it would be sort of self-healing.  For example, in cases where we enter the `if new_links_string is None:` block below, the current charm I think can never recover (that relation will never get a link, because we never try again to create one).  Whereas recomputing the entire sidebar on every event means the next event will trigger a retry of the whole thing and maybe recover. 

Also, I know we have relation events that we can act on, but I think they just make understanding the logic of the charm harder

Reply:
Implemented thanks